src/ir/possible-contents.cpp

```diff
@@ -0,0 +1,1607 @@
+      // something at index 0 already - the ExpressionLocation of this very
+      // expression, in particular), so we can use that as an indicator that we
+      // have never allocated one yet, and do so now.
+      coneReadIndex = makeSpecialLocation();
+      for (auto type : subTypes->getAllSubTypes(declaredHeapType)) {
+        connectDuringFlow(DataLocation{type, fieldIndex},
+                          SpecialLocation{coneReadIndex});
+      }
+
+      // TODO: if the old contents here were an exact type then we have an old
+      //       link here that we could remove as it is redundant (the cone
+      //       contains the exact type among the others). But removing links
+      //       is not efficient, so maybe not worth it.
+    }
+
+    // Link to the canonical location.
+    connectDuringFlow(SpecialLocation{coneReadIndex},
+                      ExpressionLocation{read, 0});
+  }
+}
+
+void Flower::writeToData(Expression* ref, Expression* value, Index fieldIndex) {
+#if defined(POSSIBLE_CONTENTS_DEBUG) && POSSIBLE_CONTENTS_DEBUG >= 2
+  std::cout << "    add special writes\n";
+#endif
+
+  // We could set up links here as we do for reads, but as we get to this code
+  // in any case, we can just flow the values forward directly. This avoids
+  // increasing the size of the graph.
+  //
+  // Figure out what to send in a simple way that does not even check whether
+  // the ref or the value was just updated: simply figure out the values being
+  // written in the current state (which is after the current update) and
+  // forward them.
+  auto refContents = getContents(getIndex(ExpressionLocation{ref, 0}));
+  auto valueContents = getContents(getIndex(ExpressionLocation{value, 0}));
+  if (refContents.isNone() || refContents.isNull()) {
+    return;
+  }
+  if (refContents.hasExactType()) {
+    // Update the one possible type here.
+    auto heapLoc =
+      DataLocation{refContents.getType().getHeapType(), fieldIndex};
+    updateContents(heapLoc, valueContents);
+  } else {
+    assert(refContents.isMany() || refContents.isGlobal());
+
+    // Update all possible subtypes here.
+    auto type = ref->type.getHeapType();
+    for (auto subType : subTypes->getAllSubTypes(type)) {
+      auto heapLoc = DataLocation{subType, fieldIndex};
+      updateContents(heapLoc, valueContents);
+    }
+  }
+}
+
+void Flower::flowRefCast(const PossibleContents& contents, RefCast* cast) {
+  // RefCast only allows valid values to go through: nulls and things of the
+  // cast type. Filter anything else out.
+  PossibleContents filtered;
+  if (contents.isMany()) {
+    // Just pass the Many through.
+    // TODO: we could emit a cone type here when we get one, instead of
+    //       emitting a Many in any of these code paths
+    filtered = contents;
+  } else {
+    auto intendedType = cast->getIntendedType();
+    bool mayBeSubType =
+      HeapType::isSubType(contents.getType().getHeapType(), intendedType);
```

Comment:
"maybe" here is because it might also be null, and it might also be simply incompatible. For example we might be trying to cast a function ref to a struct ref.

Reply:
We seem to be checking two things: 1. Is it a subtype? 2. Can it be a null?

2 is checked separately below in `if (maybeNull)`. Also if they are simply incompatible `HeapType::isSubType` will return false. So if `mayBeSubType` is true, I think there's no possible case this is not a subtype, right? Then shouldn't this be just `isSubType`?

Reply:
> We seem to be checking two things: 1. Is it a subtype? 2. Can it be a null?

Yes, we check both those things, separately. It might be a subtype but be non-nullable, or it could be a nullable type that is not a subtype, etc., so all 4 combinations can happen here.

> So if mayBeSubType is true, I think there's no possible case this is not a subtype, right? Then shouldn't this be just isSubType?

I can change it to `isSubType` if you prefer, but my reasoning was that it could also be null (which has no type, really). And I wanted to keep the two checks, of subtype and null, separate - so each of the checks is phrased as "may be X". We are only thinking about one aspect of the contents at each time. But, happy to do it either way.

Reply:
No I'm mostly asking.

- Subtype + non-nullable: `mayBeSubType` is true
- Subtype + nullable: `mayBeSubType` is true
- Not subtype + non-nullable: `mayBeSubType` is false
- Not subtype + nullable: `mayBeSubType` is false

What I meant was, is there any case when this is not a subtype when `mayBeSubType` is true? If this is a subtype whenever `mayBeSubType` is true, why is there 'maybe'?

---

However we end up, how about changing this to `maybe` (not `mayBe`)? We use 'maybe` in other places: https://github.com/WebAssembly/binaryen/blob/0b66981d0e3aa958877b5c04769e75836497b4c0/src/wasm-binary.h#L1637-L1681

Reply:
Is null a not subtype? Then, `(ref.null func)` is not a subtype of `funcref`? Didn't know that...? Anyway, sorry if this made the comment more verbose and harder to understand... I was mostly asking so please revert that if you see that is clearer.

On `maybe` vs. `mayBe`, my English is bad, sorry 😅 So nevermind.. I remembered I had seen `maybe` in other places in the code so I suggested that but I guess the meaning here was different after all.

Reply:
`(ref null func)` is a subtype of `funcref`. But the runtime value, `null`, has no type. That is, there is a single `null` for all types. So, to me at least, it seems weird to say that it must be a subtype of the heap type, if it can be a null (but it makes sense to say it's a subtype of the *type* - since that includes nullability). Anyhow, again, maybe that's a matter of definition or perspective...

Let's see what @tlively thinks. I'm fine either way here.